xbmc/guilib/GUITextLayout.h

```diff
@@ -132,14 +132,15 @@ class CGUITextLayout
                        uint32_t align);
   static void Filter(std::string &text);
 
+  void CalcTextExtent();
```

Comment:
Anytime something like this happens (moving some private member to public access we should question if this is really necessary or if we can't do this with a better approach. For an API consumer, "CalcTextExtend" doesn't ring a bell so why should it be available to the outside? Can't a better approach, using messaging be used instead?
I'll have to take a deeper look into this though as I am not familiar with this part of the code

Reply:
i can try to add the `OnMessage` callback but my concern is, how can you understand if the callback arrives before or after the same callback on GUIFontManager? What order logic is used to raise the callbacks of the messages sent?

Reply:
Don't get me wrong, I am not saying it's the way to go specially because I have no idea what I am talking about :)
I am basically saying this method is really low level and we should get to execute it some other way. Being it messaging, encapsulation, etc - basically as the consequence of some event. It's not something that you'd want to see exposed on a public api - furthermore it even retuns void as a response to "calc" something. Just seems too weird.

Corrently this method is called internally in `UpdateStyled`. Can't you simply call `UpdateStyled` on your side so that the method is internally triggered? Or wrap it into some other meaningfull method like `Refresh()` ?

Reply:
`UpdateStyled` do things that are not needed because is indeed to be used when the text change only,
here lack a way to re-update the internal cached text size when the screen scaling change (thing done by `CalcTextExtent`). 

Currently the GUITextLayout can update the chached text extent (sizes used for rendering) only when the text change and this is not so appropriate, when the screen font scaling change is kept the old values because the text is always the same, then will cause the cropped text rendering.

The `RefreshLayoutSize()` call https://github.com/xbmc/xbmc/pull/20509/files#diff-34de2075d888d6cf162eeb2df31a4901fe3636f0cbdc3acdde321e1b99134abaR78
update the text extent used to determinate the text size,
then this influence behaviour here:
https://github.com/xbmc/xbmc/pull/20509/files#diff-34de2075d888d6cf162eeb2df31a4901fe3636f0cbdc3acdde321e1b99134abaL100-L110
the calls `GetMaxWidth()` recall GUITextLayout to get the text size (now updated by RefreshLayoutSize),
and then `m_label.SetMaxRect()` call inside it, will call `UpdateRenderRect` used to update the rendering rect
also in this case pick the text size from GUITextLayout

maybe i can try use the Messaging system by addin a new type of message, then use GUIFontManager to ask to update all GUITextLayout internal size, i will have to re-check the whole chain of events because is a mess of intricate things...

Reply:
i have tried an alternative solution by using messagins at lower level so to CGUITextLayout e.g. https://github.com/CastagnaIT/xbmc/commit/5feeec4ad283008b317d2efde4eb9960d070340d
(that i find it ugly because we have to add/remove it on each control)
but this not works,
because when we update the value on `CGUITextLayout` by use OnMessage (with `CalcTextExtent`) the `m_maxRect` in `CGUILabel` will be not updated, and  `m_maxRect` value is updated from `CGUILabelControl` with `SetMaxRect` due to the custom value behaviour

if we theorise about run CalcTextExtent from `CGUILabel` and then make some kind of adaptations to update `m_maxRect` in any case `CalcTextExtent` method will be public, and i would not know how to obtain the value CGUILabelControl::GetMaxWidth from `CGUILabel`
I don't think the situation will improve much

Reply:
i have found another way, by taking inspiration from `CGUITextBox` class i have derived the control from `CGUITextLayout` then in this way i can keep `CalcTextExtent` protected as in origin.
i think now should be good see last commit added